lightning/src/chain/chainmonitor.rs

```diff
@@ -128,6 +187,37 @@ where C::Target: chain::Filter,
 		}
 	}
 
+	/// Dispatches to per-channel monitors, which are responsible for updating their on-chain view
+	/// of a channel based on transactions unconfirmed as a result of a chain reorganization. See
+	/// [`ChannelMonitor::transaction_unconfirmed`] for details.
+	///
+	/// Used instead of [`block_disconnected`] by clients that are notified of transactions rather
+	/// than blocks. May be called before or after [`update_best_block`] for transactions in the
+	/// corresponding block. See [`update_best_block`] for further calling expectations. 
+	///
+	/// [`block_disconnected`]: Self::block_disconnected
+	/// [`update_best_block`]: Self::update_best_block
+	pub fn transaction_unconfirmed(&self, txid: &Txid) {
+		let monitors = self.monitors.read().unwrap();
+		for monitor in monitors.values() {
+			monitor.transaction_unconfirmed(txid, &*self.broadcaster, &*self.fee_estimator, &*self.logger);
+		}
+	}
+
+	/// Returns the set of txids that should be monitored for confirmation on chain or for
+	/// re-organization out of the chain.
+	pub fn get_relevant_txids(&self) -> Vec<Txid> {
+		let mut txids = Vec::new();
+		let monitors = self.monitors.read().unwrap();
+		for monitor in monitors.values() {
+			txids.append(&mut monitor.get_relevant_txids());
+		}
+
+		txids.sort_unstable();
+		txids.dedup();
```

Comment:
> And we don't do transaction across monitors ? I think we already speak about it in a previous refactor.

Yes, I think that I recall a similar conversation. When you say, "we" do you mean in rust-lightning or more broadly as part of a spec requirement? I ask because we may need to handle an atypical transaction.

Reply:
The spec doesn't forbid aggregating transactions from different channels for example penalty ones or second-stage HTLCs. And we don't implement it though it could be a potential fee optimization in the future, though a bit tricky to get right.

IIRC last time we concluded we don't do it for now, but we're likely going to require dedup in the future, e.g dual-funding transaction initiating many channels.

What do you mean by an atypical transaction ?

Reply:
> The spec doesn't forbid aggregating transactions from different channels for example penalty ones or second-stage HTLCs. And we don't implement it though it could be a potential fee optimization in the future, though a bit tricky to get right.

Ah, I may have had a bit of confusion around who is generating the transactions. If we are always the one, then we won't need to dedup if we don't produce such transactions yet. But if our counterparty can also produces the transactions, then we may have to dedup if their implementation optimizes in this way. I think that's what I was trying to get at before.

> IIRC last time we concluded we don't do it for now, but we're likely going to require dedup in the future, e.g dual-funding transaction initiating many channels.

Seems like it is fine to keep the dedup if we'll eventually need it. Let me know if you prefer to drop it. It should mostly be linear in the number of channels and only involves scanning (not shifting) if we don't produce dups.

> What do you mean by an atypical transaction ?

Just as what you described above.